File: 0051-n-queens/0051-n-queens.py
class Solution:
    def solve(self, N, row, res, rf, cf, t1f, t2f, board):
        if row == N:
            ans = []
            for r in range(N):
                ans.append("".join(board[r]))
            res.append(ans)
            return

        for col in range(N):
            if rf[row] or cf[col] or t1f[row + col] or t2f[N - 1 - row + col]:
                continue

            rf[row] = 1
            cf[col] = 1
            t1f[row + col] = 1
            t2f[N - 1 - row + col] = 1
            board[row][col] = "Q"

            self.solve(N, row + 1, res, rf, cf, t1f, t2f, board)

            rf[row] = 0
            cf[col] = 0
            t1f[row + col] = 0
            t2f[N - 1 - row + col] = 0
            board[row][col] = "."

    def solveNQueens(self, N: int) -> List[List[str]]:
        res = []
        rf = [0] * N
        cf = [0] * N
        t1f = [0] * (2 * N - 1)
        t2f = [0] * (2 * N - 1)
        board=[]
        for r in range(N):
            board.append([])
            for c in range(N):
                board[r].append(".")
        self.solve(N, 0, res, rf, cf, t1f, t2f, board)
        return res

File: 0051-n-queens/0051-n-queens.java
class Solution {
    public List<List<String>> solveNQueens(int n) {
        int ned[]=new int[2*n-1];
        int swd[]=new int[2*n-1];
        int row[]=new int[n];
        int col[]=new int[n];
        List<List<String>> ans = new ArrayList<>();
        String[][] board = new String[n][n];
        for(int r=0;r<n;r++){
            for(int c=0;c<n;c++){
                board[r][c]=".";
            }
        }
        helper(ned,swd,row,col,ans,board,0,n);
        return ans;
    }
    private static void helper(int[] ned,int[] swd,int[] row,int[] col,List<List<String>> ans,String[][] board,int ind,int N){
        if(ind==N){
            List<String> temp=new ArrayList<>();
            for(int r=0;r<N;r++){
                String s="";
                for(int c=0;c<N;c++){
                    s+=board[r][c];
                }
                temp.add(s);

            }
            ans.add(temp);
            return;
        }
        for(int c=0;c<N;c++){
            
            if(ned[ind+c]!=1 && swd[c-ind+N-1]!=1 && row[ind]!=1 && col[c]!=1 && !board[ind][c].equals("Q")){
                ned[ind+c]=1;
                swd[c-ind+N-1]=1;
                row[ind]=1;
                col[c]=1;
                board[ind][c]="Q";
                helper(ned,swd,row,col,ans,board,ind+1,N);
                board[ind][c]=".";
                ned[ind+c]=0;
                swd[c-ind+N-1]=0;
                row[ind]=0;
                col[c]=0;
            }
        }
    }
}